Core/src/Material/MaterialInteractionAssignment.cpp

```diff
@@ -0,0 +1,98 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Material/MaterialInteractionAssignment.hpp"
+
+std::array<std::vector<Acts::MaterialInteraction>, 2u>
+Acts::MaterialInteractionAssignment::assign(
+    const GeometryContext& gctx,
+    const std::vector<MaterialInteraction>& materialInteractions,
+    const std::vector<std::tuple<const Surface*, Vector3, Vector3>>&
+        intersectedSurfaces,
+    const Options& options) {
+  // Assume a high assignment rate
+  std::vector<MaterialInteraction> assignedMaterialInteractions;
+  assignedMaterialInteractions.reserve(materialInteractions.size());
+
+  std::vector<MaterialInteraction> unassignedMaterialInteractions;
+
+  /// Simple matching of material interactions to surfaces - no pre/post
+  /// matching
+  // -----------------------------------------------------------------------------
+  // Double-Loop over the sorted material interactions
+  std::size_t is = 0u;
+  for (const auto& materialInteraction : materialInteractions) {
+    // First check if there is a global veto
+    bool veto = false;
+    for (const auto& gVeto : options.globalVetos) {
+      if (gVeto(materialInteraction)) {
+        unassignedMaterialInteractions.push_back(materialInteraction);
+        veto = true;
+        break;
+      }
+    }
+    // Now veto this assignment
+    if (veto) {
+      continue;
+    }
+
+    // Walk along the sorted intersections
+    auto [cSurface, cPosition, cDirection] = intersectedSurfaces[is];
+    ActsScalar cDistance = (cPosition - materialInteraction.position).norm();
+
+    // Peak forward to check if you have a closer intersection
+    while (is + 1u < intersectedSurfaces.size() &&
+           ((std::get<1>(intersectedSurfaces[is + 1]) -
+             materialInteraction.position)
+                .norm() < cDistance)) {
+      // Recalculate the new distance
+      ActsScalar nDistance = (std::get<1>(intersectedSurfaces[is + 1]) -
+                              materialInteraction.position)
+                                 .norm();
+      ++is;
+      cDistance = nDistance;
+    }
```

Comment:
It could, but we would lose the logic behind: 

- the assignment logic is given by the fact that both <material_interaction>, <surface_assignment_candidates> are orders from the origin of the mapping process.

So, what this module does is ... it walks along the material_interaction and then successively assigns them to the best suited surface_assignment_candidate. This allows me to run assign-to-next tricks (that recreate the mapping setup @Corentin-Allaire  has implemented int he original mapper).